### PR TITLE
virt-handler: Reduce Get() calls for KSM handling

### PR DESCRIPTION
**Pull request by iholder101**
### What this PR does
Partially fixes https://github.com/kubevirt/kubevirt/issues/14627.

Before this PR:
virt-handler executes a `Get()` call for the node object in every heart-beat, which is every minute.

This means that for 2k nodes, we'll do 2000 get calls a minute just for heart-beats.

After this PR:
* virt-handler will not execute `Get()` calls as part of the heart-beat process.
* virt-handler will only execute a `Get()` call if KSM needs to be handled, which is every 10 minutes or whenever Kubevirt CR's configuration change.
* whenever virt-handler would handle with KSM, it will be guaranteed to perform a single `Get()` call.

This will reduce the vast majority of `Get()` calls done by virt-handler's heartbeat.

### Release note

```release-note
virt-handler: Reduce Get() calls for KSM handling
```

### CODE REVIEW

File: pkg/virt-handler/heartbeat/ksm.go [hunk 1]
```diff
@@ -291,7 +318,12 @@ func handleKSM(node *v1.Node, clusterConfig *virtconfig.ClusterConfig) (ksmLabel
 	return true, ksm.running
 }
 
-func disableKSM(node *v1.Node) {
+func disableKSM(getNode func() (*v1.Node, error)) {
+	node, err := getNode()
+	if err != nil {
+		return
+	}
```
Comment (fossedihelm) [inline pkg/virt-handler/heartbeat/ksm.go:332]: I'd prefer to leave it as it was, leaving the `getNode` handling to the caller

Comment (iholder101) [reply pkg/virt-handler/heartbeat/ksm.go:333]: Are you sure? This will make things a lot more verbose:
```go
	if ksmConfig == nil {
		if enabled {
			node, err := getNode()
			if err != nil {
				return false, false
			}
			disableKSM(node)
		}
```

But I'm open to this approach if you think it's better.

File: pkg/virt-handler/heartbeat/ksm.go [hunk 2]
```diff
@@ -269,13 +286,24 @@ func handleKSM(node *v1.Node, clusterConfig *virtconfig.ClusterConfig) (ksmLabel
 		return false, false
 	}
 
-	if !selector.Matches(labels.Set(node.ObjectMeta.Labels)) {
-		if enabled {
-			disableKSM(node)
+	if enabled {
```
[the remark this answers is not in the document]
Comment (iholder101) [reply pkg/virt-handler/heartbeat/ksm.go:290]: Logically, it doesn't make any difference.

However, in order to check if the selector matches the node we need the node object, while we don't need to perform any `Get()` call to check whether `enabled` is true.

IOW, this makes sure we defer the `Get()` call to only whenever necessary.

[the remark this answers is not in the document]
Comment (iholder101) [reply pkg/virt-handler/heartbeat/ksm.go:290]: Looking at the code again, if `enabled` is false we will `Get()` the node anyhow.
But still perhaps the code is better written this way in case more code will be added here. WDYT?

File: pkg/virt-handler/heartbeat/heartbeat.go [hunk 3]
```diff
@@ -75,6 +78,8 @@ func (h *HeartBeat) heartBeat(heartBeatInterval time.Duration, stopCh chan struc
 	// 1 minute with a 1.2 jitter + the time it takes for the heartbeat function to run (sliding == true).
 	// So the amount of time between heartbeats randomly varies between 1min and 2min12sec + the heartbeat function execution time.
 	wait.JitterUntil(h.do, heartBeatInterval, 1.2, true, stopCh)
+
+	wait.JitterUntil(h.handleKSM, 10*time.Minute, 1.2, true, stopCh)
```
[the remark this answers is not in the document]
Comment (iholder101) [reply pkg/virt-handler/heartbeat/heartbeat.go:82]: Great catch!

File: pkg/virt-handler/heartbeat/ksm.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (iholder101) [reply pkg/virt-handler/heartbeat/ksm.go]: Nevermind, please ignore, I've reverted this.

File: pkg/virt-handler/heartbeat/ksm.go [hunk 4]
```diff
@@ -248,50 +250,88 @@ func getIntParam(node *v1.Node, param string, defaultValue, lowerBound, upperBou
 // will set the outcome value to the n.KSM struct
 // If the node labels match the selector terms, the ksm will be enabled.
 // Empty Selector will enable ksm for every node
-func handleKSM(node *v1.Node, clusterConfig *virtconfig.ClusterConfig) (ksmLabelValue, ksmEnabledByUs bool) {
+func handleKSM(nodeName string, client k8sv1.CoreV1Interface, clusterConfig *virtconfig.ClusterConfig) (ksmLabelValue, ksmEnabledByUs, needsUpdate bool) {
 	available, enabled := loadKSM()
 	if !available {
-		return false, false
+		return
```
[the remark this answers is not in the document]
Comment (iholder101) [reply pkg/virt-handler/heartbeat/ksm.go:256]: Another great catch. Addressed, thank you!

[the remark this answers is not in the document]
Comment (iholder101) [reply pkg/virt-handler/heartbeat/ksm.go:256]: See the `forceUpdate` flag that's been added.
The handler will now force update when it firstly loads, so a label will always exist. I guess we can set `needsUpdate` to true here only if we suspect that KSM could be not-available on the node then it would become available again. Is that a use-case we care about?

File: pkg/virt-handler/heartbeat/ksm.go [hunk 5]
```diff
@@ -269,13 +286,21 @@ func handleKSM(node *v1.Node, clusterConfig *virtconfig.ClusterConfig) (ksmLabel
...
+	if !selector.Matches(labels.Set(node.Labels)) {
 		if enabled {
-			disableKSM(node)
+			disableKSM(getNode)
 		}
 
 		return false, false
 	}
+
+	ksmLabelValue = true
```
Comment (fossedihelm) [inline pkg/virt-handler/heartbeat/ksm.go:304]: it's me or this is useless?

Comment (fossedihelm) [reply pkg/virt-handler/heartbeat/ksm.go:304]: ah, was just wrongly in the first commit :)

File: pkg/virt-handler/heartbeat/ksm_test.go [hunk 6]
```diff
@@ -184,7 +187,7 @@ var _ = Describe("KSM", func() {
 			fakeClient := fake.NewSimpleClientset(node)
 			heartbeat := NewHeartBeat(fakeClient.CoreV1(), deviceController(true), clusterConfig, "mynode")
 
-			heartbeat.do()
+			heartbeat.handleKSMForceUpdate()
```
Comment (fossedihelm) [inline pkg/virt-handler/heartbeat/ksm_test.go:190]: I think that here and in most cases we should test with `handleKSM()`, leaving the usage of `handleKSMForceUpdate()` to edge cases (actually it is only used once at the beginning).
Would you mind adjusting this and, if required, the node labels?

File: pkg/virt-handler/heartbeat/ksm.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (fossedihelm) [reply pkg/virt-handler/heartbeat/ksm.go:256]: I think it's good

**Review by fossedihelm — COMMENTED** [review]
Thank you!

File: pkg/virt-handler/heartbeat/ksm_test.go (hunk 6 above)
Comment (iholder101) [reply pkg/virt-handler/heartbeat/ksm_test.go]: Good point!

What I did is wrapped `newHeartbeat` in a way that it's both easier to read and write (since most of the configs are the same all over), and that executes `handleKSMForceUpdate()` once at the time of the heartbeat's creation, to simulate a first forced run.

Let me know what you think :)

Comment (fossedihelm) [reply pkg/virt-handler/heartbeat/ksm_test.go]: Makes sense, thank you! But some failures occur :)

Comment (iholder101) [reply pkg/virt-handler/heartbeat/ksm_test.go]: Fixed :)
I've tried to use `NewClientset` since `NewSimpleClientset` is deprecated, but it causes failures for some reason. Let's defer this to a follow-up :)

**Comment by fossedihelm** [comment]
/approve
Thank you!

File: pkg/virt-handler/heartbeat/ksm.go (hunk 1 above)
Comment (xpivarc) [reply pkg/virt-handler/heartbeat/ksm.go:328]: This shows that we do change the logic. Previously if `Get` for `Node` would fail we would short circuit the hearbeat, now we could fail, not disable KSM and report it is disabled?

File: pkg/virt-handler/heartbeat/ksm.go [hunk 7]
```diff
@@ -248,16 +250,31 @@ func getIntParam(node *v1.Node, param string, defaultValue, lowerBound, upperBou
...
 // If the node labels match the selector terms, the ksm will be enabled.
 // Empty Selector will enable ksm for every node
-func handleKSM(node *v1.Node, clusterConfig *virtconfig.ClusterConfig) (ksmLabelValue, ksmEnabledByUs bool) {
+func handleKSM(nodeName string, client k8sv1.CoreV1Interface, clusterConfig *virtconfig.ClusterConfig) (ksmLabelValue, ksmEnabledByUs bool) {
 	available, enabled := loadKSM()
 	if !available {
 		return false, false
 	}
 
+	var node *v1.Node
```
[the remark this answers is not in the document]
Comment (iholder101) [reply pkg/virt-handler/heartbeat/ksm.go:259]: I don't think it is. A couple of lines below, `getNode()` will always return it if non-nil, effectively caching the value and ensuring we do a single `Get()` call.

File: pkg/virt-handler/heartbeat/ksm.go (hunk 1 above)
Comment (iholder101) [reply pkg/virt-handler/heartbeat/ksm.go]: Thank you @xpivarc, you have a valid point there.

> now we could fail, not disable KSM and report it is disabled?

We could also fail with the previous code, but you're right that at the time you were writing the comment if the `Get()` call would fail then KSM would still be reported as disabled for 10 minutes.

I've now added a retry mechanism in the [last commit](https://github.com/kubevirt/kubevirt/pull/14632/commits/2f2fd5accab6e8fe1fd629d531a7a60bdc047cce), it basically runs `handleKSM()` every minute as long as it fails for any reason. However, if it succeeds, the next run would be only if kubevirt's CR config change or after 20 minutes.

WDYT?

**Comment by iholder101** [comment]
Hey @fossedihelm!

After looking at this again, I wonder why we need to perform `Get()` calls at all here.
It seems that the annotation is used to check whether the same virt-handler had previously enabled KSM or not, but the handler can check that on the node without performing an API call at all. If we need this annotation in order to report whether virt-handler had enabled KSM or not we can keep it in the patch, but still remove any `Get()` calls to the node.

Am I missing something here?

**Comment by iholder101** [comment]
> Hey @fossedihelm!
> 
> After looking at this again, I wonder why we need to perform `Get()` calls at all here. It seems that the annotation is used to check whether the same virt-handler had previously enabled KSM or not, but the handler can check that on the node without performing an API call at all. If we need this annotation in order to report whether virt-handler had enabled KSM or not we can keep it in the patch, but still remove any `Get()` calls to the node.
> 
> Am I missing something here?

I was talking to @fossedihelm offline.

Deprecating the KSM handling seems like the right move going forward from my perspective. This is a purely node-configuration task that's outside of Kubevirt's domain IMO.

However, this PR is very valuable as-is and the deprecation opens up a whole new topic that would best be discussed in a follow-up.

@xpivarc PTAL.

File: pkg/virt-handler/heartbeat/heartbeat.go [hunk 8]
```diff
@@ -78,7 +86,18 @@ func (h *HeartBeat) heartBeat(heartBeatInterval time.Duration, stopCh chan struc
...
-	wait.JitterUntil(h.handleKSM, 10*time.Minute, 0.2, true, stopCh)
+	wait.JitterUntil(func() {
+		for err := h.handleKSM(); err != nil; {
+			select {
+			case <-stopCh:
+				return
+			default:
+				log.Log.Reason(err).Errorf("failed to set KSM label on node %s, trying again in 1 minute", h.host)
+				time.Sleep(1 * time.Minute)
+				err = h.handleKSM()
```
Comment (xpivarc) [inline pkg/virt-handler/heartbeat/heartbeat.go:97]: nit: `for err := h.handleKSM(); err != nil; err =  h.handleKSM();{`

File: pkg/virt-handler/heartbeat/ksm_test.go [hunk 9]
```diff
@@ -277,40 +277,40 @@ var _ = Describe("KSM", func() {
...
 
 			By("expecting the number of pages to scan to increase every heartbeat up to max value")
-			heartbeat.handleKSM()
+			Expect(heartbeat.handleKSM()).To(Succeed())
 			expected.pages = nPagesInitDefault + pagesBoostDefault
 			expectKSMState(expected)
-			heartbeat.handleKSM()
-			heartbeat.handleKSM()
-			heartbeat.handleKSM()
+			Expect(heartbeat.handleKSM()).To(Succeed())
```
Comment (xpivarc) [inline pkg/virt-handler/heartbeat/ksm_test.go:293]: nit:  use  `for`

**Review by xpivarc — COMMENTED** [review]
Can we write unit test to test this concern https://github.com/kubevirt/kubevirt/pull/14632#discussion_r2081526575 ? 

Also I would say we should decouple this from heartbeat as it shares nothing at this point. I am ok with follow-up

File: pkg/virt-handler/heartbeat/ksm.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (iholder101) [reply pkg/virt-handler/heartbeat/ksm.go:259]: I think it worths keeping TBH. This way whoever reads the code can be confident that the object will not be fetched twice and wouldn't have to consider every possible flow. I can extract that to a dedicated type or something like that, but it seemed like an overkill. WDYT?

As a side note: after my current experience with this code, I plan to make a lot more changes there. I think that in the near future we'll be able to lose the `Get()` call entirely, or even the KSM management as a whole. But this PR will reduce 99% of the KSM related Get calls, so I prefer to leave this discussion to a follow-up.

File: pkg/virt-handler/heartbeat/heartbeat.go [hunk 10]
```diff
@@ -52,6 +52,8 @@ func NewHeartBeat(clientset k8scli.CoreV1Interface, deviceManager device_manager
 func (h *HeartBeat) Run(heartBeatInterval time.Duration, stopCh chan struct{}) (done chan struct{}) {
 	done = make(chan struct{})
 	go func() {
+		h.handleKSMForceUpdate()
```
[the remark this answers is not in the document]
Comment (iholder101) [reply pkg/virt-handler/heartbeat/heartbeat.go]: done

File: pkg/virt-handler/heartbeat/heartbeat.go (hunk 8 above)
Comment (iholder101) [reply pkg/virt-handler/heartbeat/heartbeat.go]: done

File: pkg/virt-handler/heartbeat/ksm_test.go (hunk 9 above)
Comment (iholder101) [reply pkg/virt-handler/heartbeat/ksm_test.go]: done

File: pkg/virt-handler/heartbeat/ksm.go [hunk 11]
```diff
@@ -294,26 +295,30 @@ func handleKSM(nodeName string, client k8sv1.CoreV1Interface, clusterConfig *vir
...
 	if err != nil {
 		log.DefaultLogger().Reason(err).Errorf("An error occurred while calculating the new KSM values")
-		return true, false
+		return
 	}
 
 	err = writeKsmValuesToFiles(ksm)
 	if err != nil {
 		log.DefaultLogger().Reason(err).Errorf("An error occurred while writing the new KSM values")
-		return true, false
```
[the remark this answers is not in the document]
Comment (iholder101) [reply pkg/virt-handler/heartbeat/ksm.go:288]: I strongly disagree, especially now that we have 3 return parameters of the same `bool` type.

IMHO `return true, false, false` is hard to read and write and is very easy to make mistakes with. This is especially true in our situation where we want to express "from now on, `ksmLabelValue` should be true". Simply doing `ksmLabelValue=true` is a lot easier and clear then to remember that the second argument should always be true on return expressions..

**Comment by iholder101** [comment]
@xpivarc PTAL.
Also, please note that as written above I think we eventually need to get rid of KSM handling altogether. This is mainly an intermediate step to get rid of the scale issues before we discuss the deprecation further.

Thanks for your review! :)

**Comment by iholder101** [comment]
@xpivarc I've now removed the lambda function, which I saw bugged people (and was indeed ugly), and introduced a `oneShotCache` which is a cache that ensures the value is being calculated only once. I think it looks much more elegant now.

PTAL

File: pkg/virt-handler/heartbeat/heartbeat.go [hunk 12]
```diff
@@ -52,6 +52,8 @@ func NewHeartBeat(clientset k8scli.CoreV1Interface, deviceManager device_manager
...
+		h.clusterConfig.SetConfigModifiedCallback(func() { h.handleKSM() })
```
[the remark this answers is not in the document]
Comment (xpivarc) [reply pkg/virt-handler/heartbeat/heartbeat.go]: Something like 
```

for {
  select{
  case <- clusterConfigChan:
  case <- ticker.C
  }
}

File: tools/cache/oneshot-cache.go [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+import (
+	"fmt"
+	"sync"
+)
+
+// OneShotCache is a cache that calculates its value only once.
+// It will use the provided calculation function to get the value. Once a first call to this function results
+// in a non-error value, the cache will return this value for all subsequent calls.
+type OneShotCache[T any] struct {
+	value     T
```
Comment (xpivarc) [inline tools/cache/oneshot-cache.go:31]: nit:  atomic.Value and you can drop `sync.Mutex`
nit: sync.Once instead of `isSet`

File: tools/cache/oneshot-cache.go [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+	}
+
+	return &OneShotCache[T]{
+		calcFunc:  calcFunc,
+		valueLock: sync.Mutex{},
+	}, nil
+}
+
+func (c *OneShotCache[T]) Get() (T, error) {
+	if c.isSet {
```
Comment (xpivarc) [inline tools/cache/oneshot-cache.go:49]: Don't you need to lock here?

**Comment by iholder101** [comment]
@vladikr

> I think it's a mistake to handle KSM in heartbeat, can't we simply register a handleKSM as a callback with SetConfigModifiedCallback? We can then set the label on startup and react to a callback.

I 100% agree.
I've now removed all of the KSM handling code away from the heartbeat. Instead, virt-handler's main() would force update KSM at first, then will register it as a callback. It will not reconcile KSM handling further than that.

I think this is both more correct and seriously simplifies the code.

> Alternatively, can't we turn this into a device plugin?
> 
> If it's a device plugin:
> 
> * the config activation is handled by the callback (we turn on or turn off the device plugin)
> * if kvm has been manually disabled/enabled in sysfs we track it via dp healthcheck

TBH, moving forward I think we should entirely deprecate all KSM handling, as it should not be part of kubevirt's scope. There are other tools that are dedicated for node configuration, and virt-handler does too much as it is. WDYT?

File: tools/cache/oneshot-cache.go (hunk 13 above)
Comment (iholder101) [reply tools/cache/oneshot-cache.go]: `sync.Once` is problematic here, since I want this cache to support being called multiple times if the calculation function failed and returned an error. Under the hood, `sync.Once` uses a mutex, so the current implementation is very similar to it anyhow.

I did start using an `atomic.Pointer` though, following your suggestion! Thank you!

File: tools/cache/oneshot-cache.go (hunk 14 above)
Comment (iholder101) [reply tools/cache/oneshot-cache.go]: I don't think so, since whenever this changes to `true` it'll stay this way forever.

Following your suggestion above, I'm now using `atomic.Pointer` to check for a value existence, which allowed me to drop the `isSet` field.

**Comment by iholder101** [comment]
@xpivarc @vladikr PTAL.
I think the PR is much simpler now! Thanks for your suggestions!

File: cmd/virt-handler/virt-handler.go [hunk 15]
```diff
@@ -269,6 +269,13 @@ func (app *virtHandlerApp) Run() {
 	app.clusterConfig.SetConfigModifiedCallback(app.shouldChangeLogVerbosity)
 	app.clusterConfig.SetConfigModifiedCallback(app.shouldChangeRateLimiter)
 	app.clusterConfig.SetConfigModifiedCallback(app.shouldInstallKubevirtSeccompProfile)
+	go func() {
+		forceUpdateKSM := func() { virthandler.HandleKSMUpdate(app.HostOverride, app.virtCli.CoreV1(), app.clusterConfig, true) }
+		handleKSMUpdate := func() { virthandler.HandleKSMUpdate(app.HostOverride, app.virtCli.CoreV1(), app.clusterConfig, false) }
```
Comment (xpivarc) [inline cmd/virt-handler/virt-handler.go:274]: nit: this can be outside of the goroutine

**Comment by vladikr** [comment]
Thank you very much @iholder101 ! 
It looks great to me now. 
Let's defer the discussion about a complete removal to another PR. My concern is about High performance VMs landing on nodes with enabled KSM.
/approve
/hold
Feel free to remove the hold whenever you are satisfied.

File: cmd/virt-handler/virt-handler.go (hunk 15 above)
Comment (iholder101) [reply cmd/virt-handler/virt-handler.go:274]: I tend to think it is best as is, since having this outside of the goroutine might imply that these functions are intended to be used in the following code. I generally think it's best to have these in the most inner scope possible. In addition, it has no performance impact and mainly serves as documentation with no comments (as comments are a code smell :) )

**Comment by iholder101** [comment]
> Thank you very much @iholder101 ! It looks great to me now. Let's defer the discussion about a complete removal to another PR. My concern is about High performance VMs landing on nodes with enabled KSM. /approve /hold Feel free to remove the hold whenever you are satisfied.

Sounds good!
We can discuss the follow-up to this on https://github.com/kubevirt/kubevirt/issues/14627.

/unhold
